File: 2023/day12/zig/solution.zig
const std = @import("std");

const State = struct {
    pos: u16,
    group_idx: u8,
    current_run: u8,
};

fn countArrangements(pattern: []const u8, groups: []const u8, allocator: std.mem.Allocator) !u64 {
    var memo = std.AutoHashMap(State, u64).init(allocator);
    defer memo.deinit();

    return dp(pattern, groups, 0, 0, 0, &memo);
}

fn dp(
    pattern: []const u8,
    groups: []const u8,
    pos: u16,
    group_idx: u8,
    current_run: u8,
    memo: *std.AutoHashMap(State, u64),
) u64 {
    const state = State{ .pos = pos, .group_idx = group_idx, .current_run = current_run };

    // Check memo
    if (memo.get(state)) |cached| {
        return cached;
    }

    // Base case: reached end of pattern
    if (pos == pattern.len) {
        var result: u64 = 0;
        // Valid if we've matched all groups and no partial run
        if (group_idx == groups.len and current_run == 0) {
            result = 1;
        }
        // Or if we're on the last group and the run matches
        else if (group_idx == groups.len - 1 and groups[group_idx] == current_run) {
            result = 1;
        }
        memo.put(state, result) catch {};
        return result;
    }

    var result: u64 = 0;
    const char = pattern[pos];

    // Option 1: Place operational spring (.)
    if (char == '.' or char == '?') {
        if (current_run == 0) {
            // No active run, just move forward
            result += dp(pattern, groups, pos + 1, group_idx, 0, memo);
        } else if (group_idx < groups.len and groups[group_idx] == current_run) {
            // End current run if it matches expected group size
            result += dp(pattern, groups, pos + 1, group_idx + 1, 0, memo);
        }
        // Otherwise invalid (run doesn't match group)
    }

    // Option 2: Place damaged spring (#)
    if (char == '#' or char == '?') {
        if (group_idx < groups.len and current_run < groups[group_idx]) {
            // Can extend current run
            result += dp(pattern, groups, pos + 1, group_idx, current_run + 1, memo);
        }
        // Otherwise invalid (exceeds group size or no more groups)
    }

    memo.put(state, result) catch {};
    return result;
}

fn parseLine(line: []const u8, groups_buf: []u8) !struct { pattern: []const u8, groups: []const u8 } {
    // Find the space separating pattern from groups
    var space_idx: usize = 0;
    for (line, 0..) |c, i| {
        if (c == ' ') {
            space_idx = i;
            break;
        }
    }

    const pattern = line[0..space_idx];
    const groups_str = line[space_idx + 1 ..];

    // Parse comma-separated numbers
    var group_count: usize = 0;
    var num: u8 = 0;
    var has_digit = false;

    for (groups_str) |c| {
        if (c >= '0' and c <= '9') {
            num = num * 10 + (c - '0');
            has_digit = true;
        } else if (c == ',') {
            if (has_digit) {
                groups_buf[group_count] = num;
                group_count += 1;
                num = 0;
                has_digit = false;
            }
        }
    }
    if (has_digit) {
        groups_buf[group_count] = num;
        group_count += 1;
    }

    return .{ .pattern = pattern, .groups = groups_buf[0..group_count] };
}

fn unfold(
    pattern: []const u8,
    groups: []const u8,
    unfolded_pattern_buf: []u8,
    unfolded_groups_buf: []u8,
) struct { pattern: []const u8, groups: []const u8 } {
    // Build unfolded pattern: pattern?pattern?pattern?pattern?pattern
    var pattern_len: usize = 0;
    for (0..5) |i| {
        if (i > 0) {
            unfolded_pattern_buf[pattern_len] = '?';
            pattern_len += 1;
        }
        @memcpy(unfolded_pattern_buf[pattern_len .. pattern_len + pattern.len], pattern);
        pattern_len += pattern.len;
    }

    // Build unfolded groups: groups repeated 5 times
    var groups_len: usize = 0;
    for (0..5) |_| {
        @memcpy(unfolded_groups_buf[groups_len .. groups_len + groups.len], groups);
        groups_len += groups.len;
    }

    return .{
        .pattern = unfolded_pattern_buf[0..pattern_len],
        .groups = unfolded_groups_buf[0..groups_len],
    };
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    var part1_total: u64 = 0;
    var part2_total: u64 = 0;

    // Buffers for parsing
    var groups_buf: [64]u8 = undefined;
    var unfolded_pattern_buf: [1024]u8 = undefined;
    var unfolded_groups_buf: [320]u8 = undefined;

    var lines = std.mem.splitScalar(u8, content, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        // Remove potential carriage return
        const clean_line = if (line.len > 0 and line[line.len - 1] == '\r')
            line[0 .. line.len - 1]
        else
            line;

        if (clean_line.len == 0) continue;

        const parsed = try parseLine(clean_line, &groups_buf);

        // Part 1
        const count1 = try countArrangements(parsed.pattern, parsed.groups, allocator);
        part1_total += count1;

        // Part 2: unfold and count
        const unfolded = unfold(parsed.pattern, parsed.groups, &unfolded_pattern_buf, &unfolded_groups_buf);
        const count2 = try countArrangements(unfolded.pattern, unfolded.groups, allocator);
        part2_total += count2;
    }

    std.debug.print("Part 1: {d}\n", .{part1_total});
    std.debug.print("Part 2: {d}\n", .{part2_total});
}

File: 2023/day12/c/solution.c
/*
 * Advent of Code 2023 Day 12: Hot Springs
 *
 * Uses memoized dynamic programming to count valid arrangements.
 * State: (position, group_index, current_run_length)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define MAX_PATTERN_LEN 128
#define MAX_GROUPS 64
#define MAX_RUN_LEN 32

/* Memoization table: -1 means not computed yet */
static int64_t memo[MAX_PATTERN_LEN][MAX_GROUPS][MAX_RUN_LEN];

/* Global pattern and groups for current line */
static char pattern[MAX_PATTERN_LEN];
static int groups[MAX_GROUPS];
static int pattern_len;
static int num_groups;

static void reset_memo(void) {
    memset(memo, -1, sizeof(memo));
}

/*
 * DP function: count valid arrangements starting from (pos, group_idx, current_run)
 */
static int64_t dp(int pos, int group_idx, int current_run) {
    /* Base case: reached end of pattern */
    if (pos == pattern_len) {
        /* Valid if we've matched all groups and no partial run */
        if (group_idx == num_groups && current_run == 0)
            return 1;
        /* Or if we're on the last group and the run matches */
        if (group_idx == num_groups - 1 && groups[group_idx] == current_run)
            return 1;
        return 0;
    }

    /* Check memoization */
    if (memo[pos][group_idx][current_run] != -1)
        return memo[pos][group_idx][current_run];

    int64_t result = 0;
    char c = pattern[pos];

    /* Option 1: Place operational spring (.) */
    if (c == '.' || c == '?') {
        if (current_run == 0) {
            /* No active run, just move forward */
            result += dp(pos + 1, group_idx, 0);
        } else if (group_idx < num_groups && groups[group_idx] == current_run) {
            /* End current run if it matches expected group size */
            result += dp(pos + 1, group_idx + 1, 0);
        }
        /* Otherwise invalid (run doesn't match group) */
    }

    /* Option 2: Place damaged spring (#) */
    if (c == '#' || c == '?') {
        if (group_idx < num_groups && current_run < groups[group_idx]) {
            /* Can extend current run */
            result += dp(pos + 1, group_idx, current_run + 1);
        }
        /* Otherwise invalid (exceeds group size or no more groups) */
    }

    memo[pos][group_idx][current_run] = result;
    return result;
}

static int64_t count_arrangements(void) {
    reset_memo();
    return dp(0, 0, 0);
}

/*
 * Parse a line into pattern and groups.
 * Returns 1 on success, 0 on failure.
 */
static int parse_line(const char *line) {
    /* Find the space separator */
    const char *space = strchr(line, ' ');
    if (!space)
        return 0;

    /* Copy pattern */
    int plen = (int)(space - line);
    if (plen >= MAX_PATTERN_LEN)
        return 0;
    strncpy(pattern, line, plen);
    pattern[plen] = '\0';
    pattern_len = plen;

    /* Parse groups */
    num_groups = 0;
    const char *p = space + 1;
    while (*p && *p != '\n' && *p != '\r') {
        if (num_groups >= MAX_GROUPS)
            return 0;
        groups[num_groups++] = (int)strtol(p, (char **)&p, 10);
        if (*p == ',')
            p++;
    }

    return 1;
}

/*
 * Unfold pattern and groups by repeating 5 times.
 */
static void unfold(void) {
    /* Save original values */
    char orig_pattern[MAX_PATTERN_LEN];
    int orig_groups[MAX_GROUPS];
    int orig_plen = pattern_len;
    int orig_ngroups = num_groups;

    strcpy(orig_pattern, pattern);
    memcpy(orig_groups, groups, num_groups * sizeof(int));

    /* Build unfolded pattern: pattern?pattern?pattern?pattern?pattern */
    pattern_len = 0;
    for (int i = 0; i < 5; i++) {
        if (i > 0)
            pattern[pattern_len++] = '?';
        memcpy(pattern + pattern_len, orig_pattern, orig_plen);
        pattern_len += orig_plen;
    }
    pattern[pattern_len] = '\0';

    /* Build unfolded groups */
    num_groups = 0;
    for (int i = 0; i < 5; i++) {
        memcpy(groups + num_groups, orig_groups, orig_ngroups * sizeof(int));
        num_groups += orig_ngroups;
    }
}

int main(void) {
    FILE *fp = fopen("../input.txt", "r");
    if (!fp) {
        fprintf(stderr, "Failed to open ../input.txt\n");
        return 1;
    }

    char line[256];
    int64_t part1_total = 0;
    int64_t part2_total = 0;

    while (fgets(line, sizeof(line), fp)) {
        /* Skip empty lines */
        if (line[0] == '\n' || line[0] == '\r' || line[0] == '\0')
            continue;

        if (!parse_line(line)) {
            fprintf(stderr, "Failed to parse line: %s", line);
            continue;
        }

        /* Part 1: count arrangements for original pattern */
        part1_total += count_arrangements();

        /* Part 2: unfold and count arrangements */
        unfold();
        part2_total += count_arrangements();
    }

    fclose(fp);

    printf("Part 1: %lld\n", (long long)part1_total);
    printf("Part 2: %lld\n", (long long)part2_total);

    return 0;
}
